jmhJar task is not compatible with Gradle configuration cache

**Describe the bug**
The plugin causes an error if the experimental Gradle configuration cache is enabled.

**To Reproduce**
Run `gradle jmh --configuration-cache` in a project that uses this plugin.

The output will include an error along the lines of:
```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

1 problem was found storing the configuration cache.
- Task `:project:jmhJar` of type `org.gradle.api.tasks.bundling.Jar`: cannot serialize object of type 
'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not
supported with the configuration cache.
  See https://docs.gradle.org/7.6-rc-3/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
```
This appears to happen for both shadow and non-shadow forms of the jar task.

There's an easy and obvious workaround, of course -- "don't do that" -- but it becomes a chore if the configuration cache is turned on in gradle.properties.

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.5-bin.zip
 networkTimeout=10000
 validateDistributionUrl=true
 zipStoreBase=GRADLE_USER_HOME

--- a/src/funcTest/groovy/me/champeau/jmh/AbstractFuncSpec.groovy
+++ b/src/funcTest/groovy/me/champeau/jmh/AbstractFuncSpec.groovy
@@ -26,6 +26,7 @@ abstract class AbstractFuncSpec extends Specification {
 
     protected static final List<GradleVersion> TESTED_GRADLE_VERSIONS = [
             GradleVersion.version('7.0'),
+            GradleVersion.version('8.0'),
             GradleVersion.current()
     ]
 

--- a/src/funcTest/groovy/me/champeau/jmh/ParameterSpec.groovy
+++ b/src/funcTest/groovy/me/champeau/jmh/ParameterSpec.groovy
@@ -15,7 +15,7 @@
  */
 package me.champeau.jmh
 
-import org.gradle.util.GradleVersion
+
 import spock.lang.Unroll
 
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
@@ -46,7 +46,8 @@ class ParameterSpec extends AbstractFuncSpec {
 
         then:
         result.task(":jmhJar").outcome == SUCCESS
-        result.output.contains("Calculating task graph as no configuration cache is available for tasks: jmhJar")
+        result.output.contains("Calculating task graph as no configuration cache is available for tasks: jmhJar") ||
+                result.output.contains("Calculating task graph as no cached configuration is available for tasks: jmhJar")
 
         when:
         result = build("jmhJar", "--configuration-cache")

--- a/src/funcTest/groovy/me/champeau/jmh/ParameterSpec.groovy
+++ b/src/funcTest/groovy/me/champeau/jmh/ParameterSpec.groovy
@@ -39,7 +39,6 @@ class ParameterSpec extends AbstractFuncSpec {
 
     def "executes with configuration cache"() {
         given:
-        usingGradleVersion(GradleVersion.version("7.6"))
         usingSample("java-project")
 
         when:

--- a/src/funcTest/groovy/me/champeau/jmh/UnsupportedGradleVersionSpec.groovy
+++ b/src/funcTest/groovy/me/champeau/jmh/UnsupportedGradleVersionSpec.groovy
@@ -15,7 +15,7 @@
  */
 package me.champeau.jmh
 
-import org.gradle.testkit.runner.TaskOutcome
+
 import org.gradle.util.GradleVersion
 import spock.lang.Unroll
 
@@ -32,7 +32,7 @@ class UnsupportedGradleVersionSpec extends AbstractFuncSpec {
         def result = buildAndFail("jmh")
 
         then:
-	result.output.contains('Please upgrade Gradle or use an older version of the JMH Gradle plugin.')
+	    result.output.contains('Please upgrade Gradle or use an older version of the JMH Gradle plugin.')
 
         where:
         gradleVersion << GradleVersion.version('6.9')

--- a/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy
+++ b/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy
@@ -20,6 +20,7 @@ import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.artifacts.Configuration
 import org.gradle.api.artifacts.dsl.DependencyHandler
+import org.gradle.api.file.ArchiveOperations
 import org.gradle.api.file.Directory
 import org.gradle.api.file.FileCollection
 import org.gradle.api.file.FileCopyDetails
@@ -35,10 +36,12 @@ import org.gradle.plugins.ide.eclipse.EclipseWtpPlugin
 import org.gradle.plugins.ide.idea.IdeaPlugin
 import org.gradle.util.GradleVersion
 
+import javax.inject.Inject
+
 /**
  * Configures the JMH Plugin.
  */
-class JMHPlugin implements Plugin<Project> {
+abstract class JMHPlugin implements Plugin<Project> {
     private static GradleVersion GRADLE_MIN = GradleVersion.version('7.0')
     private static boolean IS_GRADLE_MIN = GradleVersion.current() >= GRADLE_MIN
 
@@ -49,6 +52,9 @@ class JMHPlugin implements Plugin<Project> {
     static final String JMH_JAR_TASK_NAME = 'jmhJar'
     static final String JMH_TASK_COMPILE_GENERATED_CLASSES_NAME = 'jmhCompileGeneratedClasses'
     static final String JHM_RUNTIME_CLASSPATH_CONFIGURATION = 'jmhRuntimeClasspath'
+
+    @Inject
+    protected abstract ArchiveOperations getArchives()
 
     void apply(Project project) {
         assertMinimalGradleVersion()
@@ -265,6 +271,8 @@ class JMHPlugin implements Plugin<Project> {
                                                    Provider<Directory> jmhGeneratedClassesDir,
                                                    Configuration runtimeConfiguration) {
         project.tasks.register(JMH_JAR_TASK_NAME, Jar) {
+            def archives = archives
+
             it.group = JMH_GROUP
             it.dependsOn JMH_TASK_COMPILE_GENERATED_CLASSES_NAME
             it.inputs.files project.sourceSets.jmh.output
@@ -277,7 +285,7 @@ class JMHPlugin implements Plugin<Project> {
                 it.collect { it.asFile }
                         .findAll { it.directory || it.name.toLowerCase().endsWith('.jar') }
                         .collect {
-                            it.directory ? it : project.zipTree(it)
+                            it.directory ? it : archives.zipTree(it)
                         } as Set
             }).exclude(metaInfExcludes)
             def jmhSourceSetOutput = project.sourceSets.jmh.output

--- a/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy
+++ b/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy
@@ -54,7 +54,7 @@ abstract class JMHPlugin implements Plugin<Project> {
     static final String JHM_RUNTIME_CLASSPATH_CONFIGURATION = 'jmhRuntimeClasspath'
 
     @Inject
-    protected abstract ArchiveOperations getArchives()
+    protected abstract ArchiveOperations getArchiveOperations()
 
     void apply(Project project) {
         assertMinimalGradleVersion()
@@ -271,7 +271,7 @@ abstract class JMHPlugin implements Plugin<Project> {
                                                    Provider<Directory> jmhGeneratedClassesDir,
                                                    Configuration runtimeConfiguration) {
         project.tasks.register(JMH_JAR_TASK_NAME, Jar) {
-            def archives = archives
+            def archives = archiveOperations
 
             it.group = JMH_GROUP
             it.dependsOn JMH_TASK_COMPILE_GENERATED_CLASSES_NAME

--- a/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy
+++ b/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy
@@ -41,7 +41,7 @@ import javax.inject.Inject
 /**
  * Configures the JMH Plugin.
  */
-abstract class JMHPlugin implements Plugin<Project> {
+class JMHPlugin implements Plugin<Project> {
     private static GradleVersion GRADLE_MIN = GradleVersion.version('7.0')
     private static boolean IS_GRADLE_MIN = GradleVersion.current() >= GRADLE_MIN
 
@@ -52,9 +52,6 @@ abstract class JMHPlugin implements Plugin<Project> {
     static final String JMH_JAR_TASK_NAME = 'jmhJar'
     static final String JMH_TASK_COMPILE_GENERATED_CLASSES_NAME = 'jmhCompileGeneratedClasses'
     static final String JHM_RUNTIME_CLASSPATH_CONFIGURATION = 'jmhRuntimeClasspath'
-
-    @Inject
-    protected abstract ArchiveOperations getArchiveOperations()
 
     void apply(Project project) {
         assertMinimalGradleVersion()
@@ -271,8 +268,7 @@ abstract class JMHPlugin implements Plugin<Project> {
                                                    Provider<Directory> jmhGeneratedClassesDir,
                                                    Configuration runtimeConfiguration) {
         project.tasks.register(JMH_JAR_TASK_NAME, Jar) {
-            def archives = archiveOperations
-
+            def archives = project.objects.newInstance(ServiceInjection).archiveOperations
             it.group = JMH_GROUP
             it.dependsOn JMH_TASK_COMPILE_GENERATED_CLASSES_NAME
             it.inputs.files project.sourceSets.jmh.output
@@ -336,5 +332,10 @@ abstract class JMHPlugin implements Plugin<Project> {
             }
         }
         newConfig
+    }
+
+    interface ServiceInjection {
+        @Inject
+        ArchiveOperations getArchiveOperations()
     }
 }

